[velux] Add support for vane/slat position

andrewfg opened the pull request:
Fixes #12608

On the community forum, a user has requested to add support for reading and setting the vane/slat position of Venetian type blinds. See this https://community.openhab.org/t/velux-new-openhab2-binding-feedback-welcome/32926/1350

This PR does that by means of the following..
- Added a new Channel `vanePosition` of type Dimmer to the Thing type RollerShutter.
- Added support for reading and setting Functional Parameters FP1 .. FP4 and mapping to/from `vanePosition` Channel.
- Cleaned up and extended the trace logging for Functional Parameters FP1 .. FP4 values.
- Fixed compiler PMD.CompareObjectsWithEquals warnings.
- Fixed a minor compiler NPE warning.
- Added Junit test suite for new and existing functionality.
- Improved logging to enable easier tracking of state sequences via RegEx text filters.
- Many code optimisations based on learnings from the JUnit tests and the improved logging.

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

andrewfg commented:
The JAR file (for testing) is here..

[org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/8505669/org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip)

andrewfg commented:
Latest JARS are here..

1. New API version: [org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/8695644/org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip)

2. Legacy API version: [org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/8695651/org.openhab.binding.velux-3.3.0-SNAPSHOT.jar.zip)

jlaur commented:
> the CI Build on this PR now fails due to an apparent inconsistency between the new spotless formatter output (see #13027) and the (old) code format analyser.

It seems to have completed without failure now.

andrewfg commented:
> seems to have completed without failure now.

Yes. It was probably a synchronisation or caching issue.

jlaur commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/discovery/VeluxDiscoveryService.java:
```diff
@@ -68,9 +68,8 @@ public class VeluxDiscoveryService extends AbstractDiscoveryService implements R
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     private void updateLocalization() {
-        if (localization == Localization.UNKNOWN && localeProvider != null && i18nProvider != null) {
+        if (localization.equals(Localization.UNKNOWN) && (localeProvider != null) && (i18nProvider != null)) {
```
```suggestion
        if (Localization.UNKNOWN.equals(localization) && (localeProvider != null) && (i18nProvider != null)) {
```

andrewfg replied:
Ok.

jlaur commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/factory/VeluxHandlerFactory.java:
```diff
@@ -125,9 +125,8 @@ private void updateBindingState() {
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     private void updateLocalization() {
-        if (localization == Localization.UNKNOWN && localeProvider != null && i18nProvider != null) {
+        if (localization.equals(Localization.UNKNOWN) && (localeProvider != null) && (i18nProvider != null)) {
```
```suggestion
        if (Localization.UNKNOWN.equals(localization) && (localeProvider != null) && (i18nProvider != null)) {
```

andrewfg replied:
Ok.

jlaur commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java:
```diff
@@ -290,11 +291,11 @@ public enum Command {
-        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_NTF((short) 0x0508, "Error log  data from activation log."),
-        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_CFM((short) 0x0509, "Confirm lines from activation log."),
+        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_NTF((short) 0x0508, "Notification error log  data from activation log."),
+        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_CFM((short) 0x0509, "Acknowldge to confirm lines from activation log."),
```
```suggestion
        GW_GET_MULTIPLE_ACTIVATION_LOG_LINES_CFM((short) 0x0509, "Acknowledge to confirm lines from activation log."),
```

andrewfg replied:
Ok.

jlaur reviewed (changes requested):
Thanks for adding this feature. I have provided some feedback in form of one small comment per every second file. :-)

andrewfg commented on bundles/org.openhab.binding.velux/README.md:
```diff
@@ -38,6 +38,10 @@ The binding will automatically discover Velux Bridges within the local network,
 Once a Velux Bridge has been discovered, you will need to enter the `password` Configuration Parameter (see below) before the binding can communicate with it.
 And once the Velux Bridge is fully configured, the binding will automatically discover all its respective scenes and actuators (like windows and rollershutters), and place them in the Inbox.
 
+Note: When the KLF200 hub is started it provides a temporary private WiFi Access Point for initial configuration.
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/README.md:
```diff
@@ -154,6 +158,23 @@ The `position` Channel indicates the open/close state of the window (resp. rolle
+The `position`, `limitMinimum`, and `limitMaximum` are the same as described above for "window" Things.
+
+The `vanePosition` Channel only applies to Venetian blinds that have tiltable slats.
+It can only have a valid position value, if the main `position` of the Thing is fully down.
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxKLFAPI.java:
```diff
@@ -290,11 +291,11 @@ public enum Command {
         GW_GET_ACTIVATION_LOG_LINE_REQ((short) 0x0504, "Request line from activation log."),
-        GW_GET_ACTIVATION_LOG_LINE_CFM((short) 0x0505, "Confirm line from activation log."),
-        GW_ACTIVATION_LOG_UPDATED_NTF((short) 0x0506, "Confirm line from activation log."),
+        GW_GET_ACTIVATION_LOG_LINE_CFM((short) 0x0505, "Acknowldge to confirm line from activation log."),
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxProduct.java:
```diff
@@ -216,6 +228,12 @@ public String toString() {
+     * Return the product unique index.
+     * Either the serial number (for normal Velux devices), or its name (for e.g. Somfy devices).
+     *
+     * @return ..
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/ChannelVShutterPosition.java:
```diff
@@ -148,7 +147,7 @@ private ChannelVShutterPosition() {
         VeluxScene thisScene2 = thisBridgeHandler.bridgeParameters.scenes.getChannel().existingScenes
                 .get(new SceneName(sceneName));
-        if (thisScene2 == VeluxScene.UNKNOWN) {
+        if (thisScene2.equals(VeluxScene.UNKNOWN)) {
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/things/VeluxProduct.java:
```diff
@@ -385,4 +403,103 @@ public int getDisplayPosition() {
+     */
+    public boolean setFunctionalParameters(@Nullable FunctionalParameters newFunctionalParameters) {
+        if (newFunctionalParameters == null) {
+            return false;
```
No. Passing a null value of `newFunctionalParameters` is a special case. If there is no new value of functional parameters it means that nothing has been provided to override any prior value.

However I will update the Java Doc..

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/ChannelActuatorPosition.java:
```diff
@@ -138,48 +171,60 @@ private ChannelActuatorPosition() {
+                                : new VeluxProductPosition(PercentType.HUNDRED);
+                    }
+                    break;
+                default:
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:
```diff
@@ -135,12 +135,11 @@ public Thing2VeluxActuator(VeluxBridgeHandler thisBridgeHandler, ChannelUID this
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     public ProductBridgeIndex getProductBridgeIndex() {
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:
```diff
@@ -135,12 +135,11 @@ public Thing2VeluxActuator(VeluxBridgeHandler thisBridgeHandler, ChannelUID this
             mapThing2Velux();
         }
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:
```diff
@@ -152,9 +151,8 @@ public ProductBridgeIndex getProductBridgeIndex() {
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     public boolean isKnown() {
-        return (!(this.getProductBridgeIndex() == ProductBridgeIndex.UNKNOWN));
+        return (!(this.getProductBridgeIndex().equals(ProductBridgeIndex.UNKNOWN)));
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:
```diff
@@ -164,12 +162,11 @@ public boolean isKnown() {
-    @SuppressWarnings("PMD.CompareObjectsWithEquals")
     public boolean isInverted() {
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
Ok.

andrewfg commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/handler/utils/Thing2VeluxActuator.java:
```diff
@@ -164,12 +162,11 @@ public boolean isKnown() {
             mapThing2Velux();
         }
-        if (thisProduct == VeluxProduct.UNKNOWN) {
+        if (thisProduct.equals(VeluxProduct.UNKNOWN)) {
```
Ok.

andrewfg commented:
@jlaur just a status update for you..
- The user who had requested these changes has not yet finally tested and confirmed that everything is working.
- However in the meantime, I took some time to write a JUnit suite for the functions I have added.
- And based on those tests, I discovered some improvement potential, and changed the code accordingly.
- Note: the existing code is not covered by JUnit tests; that is a task for the future :)

I will send you another update when the user has confirmed the functionality.

andrewfg commented:
The latest JAR file is here..

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9048221/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

andrewfg commented:
@jlaur a further update..

The user who requested this new feature has Somfy actuators rather than Velux ones. These are notorious for being rather quirky. From the user's logs I found out that his Somfy actuators are producing undocumented State values in response to the ‘old’ API calls. Specifically, a standard Velux actuator returns State:5 when it is not moving, whereas his Somfy actuators return State:45 …

This issue is actually not problematic for the newly introduced vane position channel because I am now anyway using the ‘new’ API calls to read those values, and this API returns the State by a different means. However it is (and always was) a problem for the main position channel (on Somfy actuators).

So the PR must now also become a bug fix for the wrong State on the main position specifically onSomfy actuators, in addition to being an enhancement adding the vane position.

Bear with me...

andrewfg commented:
The latest JAR is here..

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9079300/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

andrewfg commented:
Latest JAR is here..

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9103096/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

jlaur commented:
@andrewfg, congrats with reaching 100 commits in this PR. ;)

andrewfg commented:
> congrats with reaching 100 commits in this PR

@jlaur you probably are wondering why there are so many commits, so following is an explanation of what I have been doing..

The PR was initiated by a user request from someone who has a type of shade which I myself do not physically have. So the testing cycle was quite challenging, where I would make a new JAR for the user to install, test, and send me the logs. Such a testing cycle can take many days, if not months (he was on a long vacation), and needed trials of various different hypotheses before one can hit on the right solution.

So to speed things up, I decided instead to create a JUnit test suite where I can inject data captured from the user's logs, and I can try and test very quickly many solution hypotheses, without having to go back to the user each time. This is way faster.

Furthermore, the existing code base has a 'humungous' amount of logging output, which makes it very difficult to read and discern the real functionality sequences. So I made quite a few changes in the logging in order to help me. I tried not to remove the logging of the original author since he had his own reasons for this, but I changed and added some logging so I could search the log files better using a RegEx text filter.

However...

I can now run the JUnit test suite on data from my own devices. And I could see areas where the functionality is not optimal. And I could test edge cases and potential state combinations and transitions which had not previously been checked properly. And with the better logging and RegEx filtering I can see cases where the existing code does not handle the state transitions as one would expect. And. And.

So this has resulted in me making many code improvements -- indeed over one hundred commits..  :(

Yet the good news for you, (and for me), is that once the PR is finally ready and approved by the originating user, the resulting functionality will have been tested far more stringently than it ever had been tested before. :)

andrewfg commented:
Latest JAR is here...

[org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab-addons/files/9162837/org.openhab.binding.velux-3.4.0-SNAPSHOT.jar.zip)

andrewfg commented:
@jlaur for info, the user who requested the enhancement has confirmed the functionality.

jlaur commented on bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/bridge/slip/FunctionalParameters.java:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Override
+    public boolean equals(@Nullable Object obj) {
```
When overriding `equals`, you must also override `hashCode`.

andrewfg replied:
Ok. Will do.

andrewfg replied:
Done

jlaur commented on bundles/org.openhab.binding.velux/src/test/java/org/openhab/binding/velux/test/TestNotificationsAndDatabase.java:
```diff
@@ -0,0 +1,907 @@
+     * Confirm the existing products database is initialised.
+     */
+    @Test
+    @Order(1)
```
Curious, is the order important?

andrewfg replied:
Yes. Each test leaves the test article in a given state, that is a prerequisite for the subsequent test.

jlaur reviewed (changes requested):
Thanks! One comment and one question.

jlaur reviewed (approved):
Thanks!

jlaur commented:
@andrewfg - can I persuade you into preparing a desired commit message for the squash merge? I assume you don't need all 114 commits messages concatenated, so it would be helpful with your own proposal to not mess something up.

andrewfg commented:
> desired commit message for the squash merge

I suggest the same as the PR name 'Add support for vane/slat position'
